File: unity/Assets/Scripts/Multiplayer/Leaderboard/LeaderboardPlayer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using UnityEngine;

namespace Michitai.Multiplayer.Leaderboard
{
    /// <summary>
    /// Information about a player on the leaderboard.
    /// Contains rank, player details, and custom data.
    /// Uses Unity's JsonUtility for deserialization of player data.
    /// </summary>
    /// <typeparam name="T">The type to deserialize player data into.</typeparam>
    [System.Serializable]
    public class LeaderboardPlayer<T> where T : class, new()
    {
        [SerializeField]
        private string player_data_json;     // Unity mode

        /// <summary>
        /// The player's rank on the leaderboard.
        /// </summary>
        public int rank;

        /// <summary>
        /// The player's unique ID.
        /// </summary>
        public int player_id;

        /// <summary>
        /// The player's display name.
        /// </summary>
        public string player_name;

        /// <summary>
        /// The deserialized player data object.
        /// </summary>
        public T PlayerData
        {
            get
            {
                return JsonUtility.FromJson<T>(player_data_json);
            }
        }
    }
}

File: java/src/main/java/com/michitai/multiplayer/leaderboard/LeaderboardPlayer.java
package com.michitai.multiplayer.leaderboard;

import com.fasterxml.jackson.annotation.JsonProperty;

/**
 * Information about a player on the leaderboard.
 * Contains rank, player details, and custom data.
 *
 * @param <T> The type to deserialize player data into.
 */
public class LeaderboardPlayer<T> {
    @JsonProperty("rank")
    private int rank;

    @JsonProperty("player_id")
    private int playerId;

    @JsonProperty("player_name")
    private String playerName;

    @JsonProperty("player_data_json")
    private String playerDataJson;

    @JsonProperty("player_data")
    private T playerData;

    public int getRank() {
        return rank;
    }

    public void setRank(int rank) {
        this.rank = rank;
    }

    public int getPlayerId() {
        return playerId;
    }

    public void setPlayerId(int playerId) {
        this.playerId = playerId;
    }

    public String getPlayerName() {
        return playerName;
    }

    public void setPlayerName(String playerName) {
        this.playerName = playerName;
    }

    public String getPlayerDataJson() {
        return playerDataJson;
    }

    public void setPlayerDataJson(String playerDataJson) {
        this.playerDataJson = playerDataJson;
    }

    public T getPlayerData() {
        return playerData;
    }

    public void setPlayerData(T playerData) {
        this.playerData = playerData;
    }
}
